File: spring-webflux/src/main/java/victor/training/reactive/usecase/augmenting/Augmenting.java
package victor.training.reactive.usecase.augmenting;

import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.List;


interface NameApi {

   Mono<String> fetchName(Long id);
}
interface AgeApi {

   Mono<Integer> fetchAge(Long id);
}
interface MinorApi {

   Mono<Void> notifyMinor(String name);
}
public class Augmenting {
   private NameApi nameApi;
   private AgeApi ageApi;
   private MinorApi minorApi;

   Flux<User> augment(List<Long> userIds) {
      return Flux.empty();
   }


}

class User {
   public final Long id;
   public final String name;
   public final int age;

   User(Long id, String name, int age) {
      this.id = id;
      this.name = name;
      this.age = age;
   }
}

File: spring-webflux/src/main/java/victor/training/reactive/usecase/augmenting/Augmenting.kt
package victor.training.reactive.usecase.augmenting

import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

interface NameApi {
    fun fetchName(userId: Long?): Mono<String>
}

interface AgeApi {
    fun fetchAge(userId: Long?): Mono<Int>
}

interface UnderAgeApi {
    fun notifyUnderAge(name: String?): Mono<Void?>?
}

class Augmenting(private val nameApi: NameApi, private val ageApi: AgeApi, private val underAgeApi: UnderAgeApi) {
    fun augment(userIds: List<Long?>?): Flux<User> {
        val userId = -1L
        val nameMono = nameApi!!.fetchName(userId)
        val ageMono = ageApi!!.fetchAge(userId)
        val age = 15
        if (age < 18) underAgeApi!!.notifyUnderAge("name")
        return Flux.empty()
    }
}

class User(val id: Long, val name: String, val age: Int)